File: src/main/java/org/taktik/icure/be/format/logic/impl/GenericResultFormatLogicImpl.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */


package org.taktik.icure.be.format.logic.impl;

import java.io.BufferedReader;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.Reader;
import java.io.StringReader;
import java.io.StringWriter;
import java.io.UnsupportedEncodingException;
import java.nio.charset.StandardCharsets;
import java.sql.Timestamp;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import org.apache.commons.io.IOUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.taktik.icure.dao.impl.idgenerators.UUIDGenerator;
import org.taktik.icure.entities.Contact;
import org.taktik.icure.entities.Document;
import org.taktik.icure.entities.embed.Service;
import org.taktik.icure.entities.embed.ServiceLink;
import org.taktik.icure.entities.embed.SubContact;
import org.taktik.icure.logic.FormLogic;
import org.taktik.icure.logic.HealthcarePartyLogic;
import org.xml.sax.SAXException;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;

public abstract class GenericResultFormatLogicImpl {
	protected HealthcarePartyLogic healthcarePartyLogic;
	protected UUIDGenerator uuidGen = new UUIDGenerator();
	protected FormLogic formLogic;

	@Autowired
	public void setFormLogic(FormLogic formLogic) {
		this.formLogic = formLogic;
	}

	@Autowired
	public void setHealthcarePartyLogic(HealthcarePartyLogic healthcarePartyLogic) {
		this.healthcarePartyLogic = healthcarePartyLogic;
	}

	protected void fillContactWithLines(Contact ctc, List<LaboLine> lls, String planOfActionId, String hcpId, List<String> protocolIds, List<String> formIds) {
		lls.forEach((ll) -> {
			String formId = null;
			for (int i=0;i<protocolIds.size();i++) {
				if (protocolIds.get(i).equals(ll.getRil() != null ? ll.getRil().getProtocol() : ll.getResultReference()) || protocolIds.size()==1 && protocolIds.get(i)!=null && protocolIds.get(i).startsWith("***")) {
					formId = formIds.get(i);
				}
			}

			SubContact ssc = new SubContact();
			ssc.setResponsible(hcpId);
			ssc.setDescr(ll.labo);
			ssc.setProtocol(ll.resultReference);
			ssc.setPlanOfActionId(planOfActionId);

			ssc.setStatus((ll.isResultLabResult() ? SubContact.STATUS_LABO_RESULT : SubContact.STATUS_PROTOCOL_RESULT) | SubContact.STATUS_UNREAD | (ll.ril != null && ll.ril.complete ? SubContact.STATUS_COMPLETE : 0));
			ssc.setFormId(formId);
			ssc.setServices(ll.getServices().stream().map(s -> new ServiceLink(s.getId())).collect(Collectors.toList()));

			ctc.getServices().addAll(ll.getServices());
			ctc.getSubContacts().add(ssc);
		});
	}

	protected String decodeRawData(byte[] rawData) throws IOException {
		String text;
		String frenchCp850OrCp1252 = org.taktik.icure.db.StringUtils.detectFrenchCp850Cp1252(rawData);
		if ("cp850".equals(frenchCp850OrCp1252)) {
			text = new String(rawData, "cp850");
		} else if ("cp1252".equals(frenchCp850OrCp1252)) {
			text = new String(rawData, "cp1252");
		} else {
			text = new String(rawData, StandardCharsets.UTF_8);
		}
		return text;
	}

	protected org.w3c.dom.Document getXmlDocument(Document doc) throws ParserConfigurationException, IOException, SAXException {
		DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
		DocumentBuilder dBuilder = dbFactory.newDocumentBuilder();
		return dBuilder.parse(new ByteArrayInputStream(doc.getAttachment()));
	}

	protected BufferedReader getBufferedReader(Document doc) throws IOException {
		return new BufferedReader(new StringReader(decodeRawData(doc.getAttachment())));
	}

	public static class LaboLine {
		protected String labo;
		protected String resultReference;
		protected String fullLine;

		protected List<LaboResultLine> labosList = new ArrayList<>();
		protected List<ProtocolLine> protoList = new ArrayList<>();

		protected ResultsInfosLine ril;
		protected PatientAddressLine pal;

		protected List<Service> services = new ArrayList<>();

		protected boolean resultLabResult;

		public String getLabo() {
			return labo;
		}

		public void setLabo(String labo) {
			this.labo = labo;
		}

		public String getResultReference() {
			return resultReference;
		}

		public void setResultReference(String resultReference) {
			this.resultReference = resultReference;
		}

		public String getFullLine() {
			return fullLine;
		}

		public void setFullLine(String fullLine) {
			this.fullLine = fullLine;
		}

		public List<LaboResultLine> getLabosList() {
			return labosList;
		}

		public void setLabosList(List<LaboResultLine> labosList) {
			this.labosList = labosList;
		}

		public List<ProtocolLine> getProtoList() {
			return protoList;
		}

		public void setProtoList(List<ProtocolLine> protoList) {
			this.protoList = protoList;
		}

		public ResultsInfosLine getRil() {
			return ril;
		}

		public void setRil(ResultsInfosLine ril) {
			this.ril = ril;
		}

		public PatientAddressLine getPal() {
			return pal;
		}

		public void setPal(PatientAddressLine pal) {
			this.pal = pal;
		}

		public List<Service> getServices() {
			return services;
		}

		public void setServices(List<Service> services) {
			this.services = services;
		}

		public boolean isResultLabResult() {
			return resultLabResult;
		}

		public void setResultLabResult(boolean resultLabResult) {
			this.resultLabResult = resultLabResult;
		}
	}

	class PatientLine {
		protected String lastName;
		protected String firstName;
		protected Timestamp dn;
		protected String sex;
		protected String protocol;

		public String getLastName() {
			return lastName;
		}

		public void setLastName(String lastName) {
			this.lastName = lastName;
		}

		public String getFirstName() {
			return firstName;
		}

		public void setFirstName(String firstName) {
			this.firstName = firstName;
		}

		public Timestamp getDn() {
			return dn;
		}

		public void setDn(Timestamp dn) {
			this.dn = dn;
		}

		public String getSex() {
			return sex;
		}

		public void setSex(String sex) {
			this.sex = sex;
		}

		public String getProtocol() {
			return protocol;
		}

		public void setProtocol(String protocol) {
			this.protocol = protocol;
		}
	}

	class PatientAddressLine {
		protected String protocol;
		protected String address;
		protected String number;
		protected String zipCode;
		protected String locality;

		public String getProtocol() {
			return protocol;
		}

		public void setProtocol(String protocol) {
			this.protocol = protocol;
		}

		public String getAddress() {
			return address;
		}

		public void setAddress(String address) {
			this.address = address;
		}

		public String getNumber() {
			return number;
		}

		public void setNumber(String number) {
			this.number = number;
		}

		public String getZipCode() {
			return zipCode;
		}

		public void setZipCode(String zipCode) {
			this.zipCode = zipCode;
		}

		public String getLocality() {
			return locality;
		}

		public void setLocality(String locality) {
			this.locality = locality;
		}
	}

	class PatientSSINLine {
		protected String protocol;
		protected String ssin;

		public String getProtocol() {
			return protocol;
		}

		public void setProtocol(String protocol) {
			this.protocol = protocol;
		}

		public String getSsin() {
			return ssin;
		}

		public void setSsin(String ssin) {
			this.ssin = ssin;
		}
	}

	class ProtocolLine {
		protected String protocol;
		protected String code;
		protected String text;

		public String getProtocol() {
			return protocol;
		}

		public void setProtocol(String protocol) {
			this.protocol = protocol;
		}

		public String getCode() {
			return code;
		}

		public void setCode(String code) {
			this.code = code;
		}

		public String getText() {
			return text;
		}

		public void setText(String text) {
			this.text = text;
		}
	}

	class LaboResultLine {
		protected String protocol;
		protected String analysisCode;
		protected String analysisType;
		protected String referenceValues;
		protected String unit;
		protected String severity;
		protected String value;

		public String getProtocol() {
			return protocol;
		}

		public void setProtocol(String protocol) {
			this.protocol = protocol;
		}

		public String getAnalysisCode() {
			return analysisCode;
		}

		public void setAnalysisCode(String analysisCode) {
			this.analysisCode = analysisCode;
		}

		public String getAnalysisType() {
			return analysisType;
		}

		public void setAnalysisType(String analysisType) {
			this.analysisType = analysisType;
		}

		public String getReferenceValues() {
			return referenceValues;
		}

		public void setReferenceValues(String referenceValues) {
			this.referenceValues = referenceValues;
		}

		public String getUnit() {
			return unit;
		}

		public void setUnit(String unit) {
			this.unit = unit;
		}

		public String getSeverity() {
			return severity;
		}

		public void setSeverity(String severity) {
			this.severity = severity;
		}

		public String getValue() {
			return value;
		}

		public void setValue(String value) {
			this.value = value;
		}
	}

	class ResultsInfosLine {
		protected String protocol;
		protected Instant demandDate;
		protected boolean complete;

		public String getProtocol() {
			return protocol;
		}

		public void setProtocol(String protocol) {
			this.protocol = protocol;
		}

		public Instant getDemandDate() {
			return demandDate;
		}

		public void setDemandDate(Instant demandDate) {
			this.demandDate = demandDate;
		}

		public boolean isComplete() {
			return complete;
		}

		public void setComplete(boolean complete) {
			this.complete = complete;
		}
	}

	class Reference {
		protected Double minValue;
		protected Double maxValue;
		protected String unit;

		public Double getMinValue() {
			return minValue;
		}

		public void setMinValue(Double minValue) {
			this.minValue = minValue;
		}

		public Double getMaxValue() {
			return maxValue;
		}

		public void setMaxValue(Double maxValue) {
			this.maxValue = maxValue;
		}

		public String getUnit() {
			return unit;
		}

		public void setUnit(String unit) {
			this.unit = unit;
		}
	}
}

File: src/main/kotlin/org/taktik/icure/be/format/logic/impl/GenericResultFormatLogicImpl.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.be.format.logic.impl

import kotlinx.coroutines.flow.Flow
import org.springframework.core.io.buffer.DataBuffer
import org.taktik.icure.asynclogic.FormLogic
import org.taktik.icure.asynclogic.HealthcarePartyLogic
import org.taktik.icure.be.format.logic.ResultFormatLogic
import org.taktik.couchdb.id.UUIDGenerator
import org.taktik.icure.db.StringUtils.detectFrenchCp850Cp1252
import org.taktik.icure.entities.Document
import org.taktik.icure.entities.HealthcareParty
import org.taktik.icure.entities.Patient
import org.taktik.icure.entities.embed.Service
import org.taktik.icure.entities.embed.ServiceLink
import org.taktik.icure.entities.embed.SubContact
import org.xml.sax.SAXException
import java.io.BufferedReader
import java.io.ByteArrayInputStream
import java.io.IOException
import java.io.StringReader
import java.nio.ByteBuffer
import java.nio.charset.CharacterCodingException
import java.nio.charset.Charset
import java.nio.charset.StandardCharsets
import java.sql.Timestamp
import java.time.Instant
import java.time.LocalDateTime
import java.util.*
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.parsers.ParserConfigurationException

abstract class GenericResultFormatLogicImpl(val healthcarePartyLogic: HealthcarePartyLogic, val formLogic: FormLogic) : ResultFormatLogic {
    protected var uuidGen = UUIDGenerator()

    override fun doExport(sender: HealthcareParty?, recipient: HealthcareParty?, patient: Patient?, date: LocalDateTime?, ref: String?, mimeType: String?, content: ByteArray?) : Flow<DataBuffer> {
        throw UnsupportedOperationException("Not implemented")
    }

    protected fun fillContactWithLines(lls: List<LaboLine>, planOfActionId: String?, hcpId: String?,
                                       protocolIds: List<String?>, formIds: List<String?>) =
        lls.map { ll: LaboLine ->
            var formId: String? = null
            for (i in protocolIds.indices) {
                if ((protocolIds[i] == ll.ril?.protocol ?: ll.resultReference) || protocolIds.size == 1 && protocolIds[i]?.startsWith("***") == true) {
                    formId = formIds[i]
                }
            }
            SubContact(
                    id = uuidGen.newGUID().toString(),
                    responsible = hcpId,
                    descr = ll.labo,
                    protocol = ll.resultReference,
                    planOfActionId = planOfActionId,
                    status = ((if (ll.isResultLabResult) SubContact.STATUS_LABO_RESULT else SubContact.STATUS_PROTOCOL_RESULT)
                            or SubContact.STATUS_UNREAD or if (ll.ril != null && ll.ril!!.isComplete) SubContact.STATUS_COMPLETE else 0),
                    formId = formId,
                    services = ll.services.map { s: Service -> ServiceLink(s.id) }.toList()
            ) to ll.services
        }


    @Throws(IOException::class)
    protected fun decodeRawData(rawData: ByteArray?): String? {
        var text: String? = null
        // Test BOM
// Test utf-16 byte order mark presence
        if (rawData != null) {
            val utf8Decoder = StandardCharsets.UTF_8.newDecoder()
            text = try {
                val decodedChars = utf8Decoder.decode(ByteBuffer.wrap(rawData))
                decodedChars.toString()
            } catch (e: CharacterCodingException) {
                val frenchCp850OrCp1252 = detectFrenchCp850Cp1252(rawData)
                val charset = if ("cp850" == frenchCp850OrCp1252) Charset.forName("cp850") else Charset.forName("cp1252")
                String(rawData, charset)
            }
        }
        return text
    }

    @Throws(ParserConfigurationException::class, IOException::class, SAXException::class)
    protected fun getXmlDocument(doc: Document, enckeys: List<String?>?): org.w3c.dom.Document {
        val dbFactory = DocumentBuilderFactory.newInstance()
        val dBuilder = dbFactory.newDocumentBuilder()
        return dBuilder.parse(ByteArrayInputStream(doc.decryptAttachment(enckeys)))
    }

    @Throws(IOException::class)
    protected fun getBufferedReader(doc: Document, enckeys: List<String?>?): BufferedReader? {
        return decodeRawData(doc.decryptAttachment(enckeys))?. let { BufferedReader(StringReader(it)) }
    }

    class LaboLine {
        var labo: String? = null
        var resultReference: String? = null
        var fullLine: String? = null
        var labosList: MutableList<LaboResultLine?> = ArrayList()
        var protoList: MutableList<ProtocolLine?> = ArrayList()
        var ril: ResultsInfosLine? = null
        var pal: PatientAddressLine? = null
        var services: MutableList<Service> = ArrayList()
        var isResultLabResult = false
    }

    inner class PatientLine {
        var lastName: String? = null
        var firstName: String? = null
        var dn: Timestamp? = null
        var sex: String? = null
        var protocol: String? = null

    }

    inner class PatientAddressLine {
        var protocol: String? = null
        var address: String? = null
        var number: String? = null
        var zipCode: String? = null
        var locality: String? = null

    }

    inner class PatientSSINLine {
        var protocol: String? = null
        var ssin: String? = null

    }

    inner class ProtocolLine {
        var protocol: String? = null
        var code: String? = null
        var text: String? = null

    }

    inner class LaboResultLine {
        var protocol: String? = null
        var analysisCode: String? = null
        var analysisType: String? = null
        var referenceValues: String? = null
        var unit: String? = null
        var severity: String? = null
        var value: String? = null

    }

    inner class ResultsInfosLine {
        var protocol: String? = null
        var demandDate: Instant? = null
        var isComplete = false

    }

    inner class Reference {
        var minValue: Double? = null
        var maxValue: Double? = null
        var unit: String? = null

    }
}
